### ISSUE TITLE
Allow parent to work on relative paths as well

### ISSUE THREAD

**Issue by harendra-kumar** [commercialhaskell/path#95]
Once we merge #34, then parent automatically works properly on relative paths as well, we just need to change the signature to:

```haskell
parent :: Path b t -> Path b Dir
```

Semantics:

```
parent "x" = "."
parent "x/y" = "x"
```

**Comment by NorfairKing** [comment]
👍

**Comment by sjakobi** [comment]
And `parent "." = "."`? I don't see how this can be solved well.

**Comment by NorfairKing** [comment]
> And parent "." = "."?

That does follow the `parent "/" = "/"` convention...

**Comment by harendra-kumar** [comment]
If we think of "." as RelRoot it is consistent.

Note that when we say `parent`  we mean the parent in the given path component and not the parent in the file system, we have no knowledge of the file system. We should not confuse the parent of "." the path with the parent of the current directory in the file system, that is the realm of path resolution to convert from relative paths to absolute. Here, we are only dealing with pure paths and "." is a concept of the last path component and not the file system's view of the current directory. 

This is another problem because of the overloading of the meaning of ".". We can create a small writeup to clear up the concepts and separate the file system view from the pure path view of things.

**Comment by harendra-kumar** [comment]
BTW, POSIX `dirname` the equivalent of our `parent` API has the same behavior:

```
$ dirname .
.
```

Though it sounds more intuitive with the API named `dirname` because it does not sound like we are resolving the parent of "." in the file system hierarchy.

**Comment by NorfairKing** [comment]
Has there been any progress on this matter? Does anyone else have an opinion on the whole matter, not just this issue, but all the relevant ones. @sjakobi @harendra-kumar @chrisdone @mrkkrp @joehillen ...

**Comment by harendra-kumar** [comment]
The first step is to understand, approve and merge #34. If that itself is not accepted then this issue is moot.

**Comment by mrkkrp** [comment]
Now that #34 is merged, would you like to prepare a PR for this to be included in the version 0.6.0?
